sw/device/lib/dif/dif_otp_ctrl.c

```diff
@@ -3,3 +3,856 @@
+      reg1 = OTP_CTRL_OWNER_SW_CFG_DIGEST_1_REG_OFFSET;
+      break;
+    case kDifOtpCtrlPartitionHwCfg:
+      reg0 = OTP_CTRL_HW_CFG_DIGEST_0_REG_OFFSET;
+      reg1 = OTP_CTRL_HW_CFG_DIGEST_1_REG_OFFSET;
+      break;
+    case kDifOtpCtrlPartitionSecret0:
+      reg0 = OTP_CTRL_SECRET0_DIGEST_0_REG_OFFSET;
+      reg1 = OTP_CTRL_SECRET0_DIGEST_1_REG_OFFSET;
+      break;
+    case kDifOtpCtrlPartitionSecret1:
+      reg0 = OTP_CTRL_SECRET1_DIGEST_0_REG_OFFSET;
+      reg1 = OTP_CTRL_SECRET1_DIGEST_1_REG_OFFSET;
+      break;
+    case kDifOtpCtrlPartitionSecret2:
+      reg0 = OTP_CTRL_SECRET2_DIGEST_0_REG_OFFSET;
+      reg1 = OTP_CTRL_SECRET2_DIGEST_1_REG_OFFSET;
+      break;
+    default:
+      return kDifOtpCtrlDigestBadArg;
+  }
+
+  uint64_t value = mmio_region_read32(otp->params.base_addr, reg1);
+  value <<= 32;
+  value |= mmio_region_read32(otp->params.base_addr, reg0);
+
+  if (value == 0) {
+    return kDifOtpCtrlDigestMissing;
+  }
+  *digest = value;
+
+  return kDifOtpCtrlDigestOk;
+}
+
+dif_otp_ctrl_dai_result_t dif_otp_ctrl_read_blocking(
+    const dif_otp_ctrl_t *otp, dif_otp_ctrl_partition_t partition,
+    uint32_t address, uint32_t *buf, size_t len) {
+  if (otp == NULL || partition >= ARRAYSIZE(kPartitions) || buf == NULL) {
+    return kDifOtpCtrlDaiBadArg;
+  }
+
+  if (!kPartitions[partition].is_software) {
+    return kDifOtpCtrlDaiBadPartition;
+  }
+
+  if ((address & kPartitions[partition].align_mask) != 0) {
+    return kDifOtpCtrlDaiUnaligned;
+  }
+
+  if (address + len >= kPartitions[partition].len) {
+    return kDifOtpCtrlDaiOutOfRange;
+  }
+
+  ptrdiff_t reg_offset = OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET +
+                         kPartitions[partition].start_addr + address;
+  mmio_region_memcpy_from_mmio32(otp->params.base_addr, reg_offset, buf,
+                                 len * sizeof(uint32_t));
+  return kDifOtpCtrlDaiOk;
+}
+
+dif_otp_ctrl_result_t dif_otp_ctrl_read_test(const dif_otp_ctrl_t *otp,
+                                             uint32_t address, uint32_t *buf,
+                                             size_t len) {
+  if (otp == NULL || buf == NULL) {
+    return kDifOtpCtrlBadArg;
+  }
+
+  ptrdiff_t reg_offset = OTP_CTRL_TEST_ACCESS_REG_OFFSET + address;
+  mmio_region_memcpy_from_mmio32(otp->params.base_addr, reg_offset, buf,
+                                 len * sizeof(uint32_t));
+  return kDifOtpCtrlOk;
+}
+
+dif_otp_ctrl_result_t dif_otp_ctrl_write_test(const dif_otp_ctrl_t *otp,
+                                              uint32_t address,
+                                              const uint32_t *buf, size_t len) {
+  if (otp == NULL || buf == NULL) {
+    return kDifOtpCtrlBadArg;
+  }
+
+  ptrdiff_t reg_offset = OTP_CTRL_TEST_ACCESS_REG_OFFSET + address;
+  mmio_region_memcpy_to_mmio32(otp->params.base_addr, reg_offset, buf,
+                               len * sizeof(uint32_t));
+  return kDifOtpCtrlOk;
+}
```

Comment:
nit: I know these are test functions but a bounds check would be nice.

Reply:
There's no bounds check because the bounds are undocumented. =/